### PR TITLE
[WIP] catch floating number divide by zero

### PR DESCRIPTION
**Pull request by apeforest**
## Description ##
NaN errors could be due to divide-by-zero floating number in C++ backend. It is not caught by default. 
Turnining on the signal exception can report such errors when running python unit test.
However, this exception currently only works for Linux OS. Need more feedbacks on a better handling cross platforms.

### CODE REVIEW

File: src/engine/engine.cc [hunk 1]
```diff
@@ -35,6 +55,18 @@ inline Engine* CreateEngine() {
   if (type == nullptr) type = "ThreadedEnginePerDevice";
   std::string stype = type;
 
+#ifdef _TEST_FLOAT_EXCEPTION
+  struct sigaction act;
+  memset(&act, '\0', sizeof(act));
+  act.sa_sigaction = &float_num_err_handler;
+  act.sa_flags = SA_SIGINFO;
+  if (sigaction(SIGFPE, &act, NULL) < 0) {
+    fprintf(stderr, "Cannot register signal handler\n");
```
Comment (yuxihu) [inline src/engine/engine.cc]: Might be also good to log the errno in the message such that we can use it to see why it failed.

File: src/engine/engine.cc [hunk 2]
```diff
@@ -23,10 +23,30 @@
  * \brief Implementation of engine.
  */
 #include <mxnet/engine.h>
+#include <fenv.h>
```
Comment (yuxihu) [inline src/engine/engine.cc]: QQ: do we follow any rules (like alphabetic order of library name) to include libraries?

**Comment by harshp8l** [comment]
@mxnet-label-bot add [pr-work-in-progress]

**Comment by stu1130** [comment]
@apeforest could you address the comments of @yuxihu Thanks!

File: src/engine/engine.cc [hunk 3]
```diff
@@ -23,10 +23,30 @@
...
+    fprintf(stderr, ">>> Integer division by zero\n");
+    break;
+  case FPE_FLTOVF:
+    fprintf(stderr, ">>> Floating overflow trap\n");
+    break;
+  case FPE_FLTDIV:
+    fprintf(stderr, ">>> Floating/decimal division by zero\n");
+    break;
+  }
+  assert(0);
```
Comment (stu1130) [inline src/engine/engine.cc]: why adding assert(0) here?

**Comment by roywei** [comment]
@apeforest ping for update. Thanks!

**Comment by sandeep-krishnamurthy** [comment]
@anirudh2290 - FYI

**Comment by pinaraws** [comment]
@apeforest Any updates?

**Comment by apeforest** [comment]
Don't have time to work on this now. Closing it.

**Pull request closed without merging**